exciting_environments/mass_spring_damper/mass_spring_damper_env.py

```diff
@@ -15,121 +18,179 @@ class MassSpringDamper(core_env.CoreEnvironment):
             d_deflection = velocity
-            d_y = d_deflection, d_velocity[0]
+            d_y = d_deflection, d_velocity  # [0]
             return d_y
 
         term = diffrax.ODETerm(vector_field)
         t0 = 0
         t1 = self.tau
-        y0 = tuple(env_states)
+        y0 = tuple([env_states.deflection, env_states.velocity])
         env_state = self._solver.init(term, t0, t1, y0, args)
         y, _, _, env_state, _ = self._solver.step(
             term, t0, t1, y0, args, env_state, made_jump=False)
 
         deflection_k1 = y[0]
         velocity_k1 = y[1]
 
-        env_states_k1 = jnp.hstack((
-            deflection_k1,
-            velocity_k1,
-        ))
-        env_states_k1_norm = env_states_k1/env_state_normalizer
+        phys = self.PhysicalStates(
+            deflection=deflection_k1, velocity=velocity_k1)
+        opt = None  # Optional(something=...)
+        return self.States(physical_states=phys, PRNGKey=None, optional=None)
 
-        return env_states_k1_norm
+    @partial(jax.jit, static_argnums=0)
+    def init_states(self):
+        """Returns default initial states for all batches."""
+        phys = self.PhysicalStates(deflection=jnp.zeros(
+            self.batch_size), velocity=jnp.zeros(self.batch_size))
+        opt = None  # self.Optional(something=jnp.zeros(self.batch_size))
+        return self.States(physical_states=phys, PRNGKey=None, optional=opt)
 
     @partial(jax.jit, static_argnums=0)
-    def default_reward_func(self, obs, action):
-        return ((obs[0])**2 + 0.1*(obs[1])**2 + 0.1*(action[0])**2)
+    def default_reward_func(self, obs, action, action_constraints):
+        """Returns reward for one batch."""
+        reward = (obs[0])**2 + 0.1*(obs[1])**2 + 0.1 * \
+            (action[0]/action_constraints.force)**2
+        return jnp.array([reward])
 
     @partial(jax.jit, static_argnums=0)
-    def generate_observation(self, states):
-        """Returns states."""
-        return states
+    def generate_observation(self, states, physical_constraints):
+        """Returns observation for one batch."""
+        obs = jnp.hstack((
+            states.physical_states.deflection / physical_constraints.deflection,
+            states.physical_states.velocity / physical_constraints.velocity,
+        ))
+        return obs
+
+    @property
+    def obs_description(self):
+        return np.array(["deflection", "velocity"])
 
     @partial(jax.jit, static_argnums=0)
-    def generate_truncated(self, states):
-        """Returns states."""
-        return jnp.abs(states) > 1
+    def generate_truncated(self, states, physical_constraints):
+        """Returns truncated information for one batch."""
+        _states = jnp.hstack((
+            states.physical_states.deflection / physical_constraints.deflection,
+            states.physical_states.velocity / physical_constraints.velocity,
+        ))
+        return jnp.abs(_states) > 1
```

Comment:
Isn't this the same as 

```suggestion
        _obs = self.generate_observation(states, physical_constraints)
        return jnp.abs(_obs) > 1
```

A bit clearer and code reuse is (almost) always good